### PR TITLE
[RLlib; docs] RLlib docs redo: New API stack Episodes (SingleAgentEpisode).

### PR DESCRIPTION
**Pull request by sven1977**
RLlib docs redo: New API stack Episodes (SingleAgentEpisode).

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/rllib/images/episodes/sa_episode_cut_and_lookback.svg
Comment (simonsays1980) [inline doc/source/rllib/images/episodes/sa_episode_cut_and_lookback.svg:1]: Awesome diagrams. Makes the lookback buffer cristal clear! 

Only thing: Describe the length of the lookback in this case :)

File: doc/source/rllib/images/episodes/sa_episode_getters.svg
Comment (simonsays1980) [inline doc/source/rllib/images/episodes/sa_episode_getters.svg:1]: Awesome!

File: doc/source/rllib/images/episodes/usage_of_episodes.svg
Comment (simonsays1980) [inline doc/source/rllib/images/episodes/usage_of_episodes.svg:1]: Really nice! I think this is well understandable.

File: doc/source/rllib/single-agent-episode.rst [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+
+Episodes
+========
+
+RLlib stores and transports all trajectory data in the form of `Episodes`, in particular
+:py:class:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode` for single-agent setups
+and :py:class:`~ray.rllib.env.multi_agent_episode.MultiAgentEpisode` for multi-agent setups.
+The data is translated from this `Episode` format to tensor batches (including a possible move to the GPU)
+only immediately before a neural network forward pass.
```
Comment (simonsays1980) [inline doc/source/rllib/single-agent-episode.rst:15]: Maybe we add here that this happens inside of the connector pipeline?

File: doc/source/rllib/single-agent-episode.rst [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+.. figure:: images/episodes/usage_of_episodes.svg
+    :width: 750
+    :align: left
+
+    **Episodes** are the main vehicle to store and transport trajectory data across the different components
+    of RLlib (for example from `EnvRunner` to `Learner` or from `ReplayBuffer` to `Learner`).
+    One of the main design principles of RLlib's new API stack is that all trajectory data is kept in such episodic form
+    for as long as possible. Only immediately before the neural network passes, "connectors" translate lists of Episodes into tensor
```
Comment (simonsays1980) [inline doc/source/rllib/single-agent-episode.rst:24]: Ah here they come.

File: doc/source/rllib/single-agent-episode.rst [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+Now that there is a :py:class:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode` to work with, one can explore
+and extract information from this episode using its different "getter" methods:
+
+.. figure:: images/episodes/sa_episode_getters.svg
+    :width: 750
+    :align: left
+
+    **SingleAgentEpisode getter APIs**: "getter" methods exist for all of the Episode's fields, which are `observations`,
+    `actions`, `rewards`, `infos`, and `extra_model_outputs`. For simplicity, only the getters for observations, actions, and rewards
```
Comment (simonsays1980) [inline doc/source/rllib/single-agent-episode.rst:90]: We should mention that the getter of `extra_model_outputs`  works a bit differently - this is specifically useful to know, if all extra-model-outputs should be extracted for a single timestep (we have then to iterate through the keys)

File: doc/source/rllib/single-agent-episode.rst [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    :start-after: rllib-sa-episode-02-begin
+    :end-before: rllib-sa-episode-02-end
+
+
+Finalized and Non-Finalized Episodes
+------------------------------------
+
+The data in a :py:class:`~ray.rllib.env.single_agent_episode.SingleAgentEpisode` can exist in two states:
+non-finalized and finalized. A non-finalized episode stores its observations and other data items in
+plain python lists and appends new timestep data to these. In a finalized episode,
```
Comment (simonsays1980) [inline doc/source/rllib/single-agent-episode.rst:108]: Unsure: to these -> to them?

File: doc/source/rllib/single-agent-episode.rst [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+state (data stored in python lists), making it very fast to append data from an ongoing episode:
+
+
+.. literalinclude:: doc_code/sa_episode.py
+    :language: python
+    :start-after: rllib-sa-episode-03-begin
+    :end-before: rllib-sa-episode-03-end
+
+
+To illustrate the differences between the data stored in a non-finalized episode vs the same data stored in
```
Comment (simonsays1980) [inline doc/source/rllib/single-agent-episode.rst:129]: vs -> vs. :)

File: doc/source/rllib/single-agent-episode.rst [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+a finalized one, take a look at this complex observation example here, showing the exact same observation data in two
+episodes (one not finalized the other finalized):
```
Comment (simonsays1980) [inline doc/source/rllib/single-agent-episode.rst:131]: not finalized -> non-finalized?

File: doc/source/rllib/single-agent-episode.rst [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    **Complex observations in a non-finalized episode**: Each individual observation is a (complex) dict matching the
+    gym environment's observation space. There are three such observation items stored in the episode so far.
+
+.. figure:: images/episodes/sa_episode_finalized.svg
+    :width: 600
+    :align: left
+
+    **Complex observations in a finalized episode**: The entire observation record is a single (complex) dict matching the
+    gym environment's observation space. At the leafs of the structure are `np.NDArrays` holding the individual values of the leaf.
+    Note that these `NDArrays` have an extra batch dim (axis=0), whose length matches the length of the episode stored (here 3).
```
Comment (simonsays1980) [inline doc/source/rllib/single-agent-episode.rst:146]: Either both `np.NDArray`  or both `NDArray` I'd say. What do you think?

File: doc/source/rllib/single-agent-episode.rst [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+action computing forward pass:
+
+.. literalinclude:: doc_code/sa_episode.py
+    :language: python
+    :start-after: rllib-sa-episode-05-begin
+    :end-before: rllib-sa-episode-05-end
+
+Another useful getter argument (besides `fill`) is the `neg_index_as_lookback` boolean argument.
+If set to True, negative indices are not interpreted as "from the end", but as
+"into the lookback buffer". This allows you to loop over a range of global timesteps
```
Comment (simonsays1980) [inline doc/source/rllib/single-agent-episode.rst:208]: This is a tiny tiny bit ambiguous when the "lookback" is actually not into the "lookback buffer" but just backwards from the considered global timestep. So, precisely it would be "from the global timestep (if needed into the lookback buffer)", wouldn't it?

File: rllib/env/utils/infinite_lookback_buffer.py [hunk 9]
```diff
@@ -565,10 +565,10 @@ def _interpret_slice(
             stop = len_self_plus_lookback
         # Stop is negative.
         elif stop < 0:
-            # `neg_indices_left_of_zero=True` -> User wants to index into the lookback
+            # `neg_index_as_lookback=True` -> User wants to index into the lookback
             # range. Set to 0 (beginning of lookback buffer) if result is a negative
             # index.
-            if neg_indices_left_of_zero:
+            if neg_index_as_lookback:
```
Comment (simonsays1980) [inline rllib/env/utils/infinite_lookback_buffer.py:571]: Nice change! This is smoother and sounds more correct :)

**Review by simonsays1980 — APPROVED** [review]
LGTM. We are moving forward here. Very nice docs for the complex episodes. Amazing how simpel and clear this is described @sven1977!!

File: doc/source/rllib/images/episodes/sa_episode_cut_and_lookback.svg
Comment (sven1977) [reply doc/source/rllib/images/episodes/sa_episode_cut_and_lookback.svg:1]: fixed

File: doc/source/rllib/single-agent-episode.rst (hunk 1 above)
Comment (sven1977) [reply doc/source/rllib/single-agent-episode.rst:18]: done

File: doc/source/rllib/single-agent-episode.rst (hunk 2 above)
Comment (sven1977) [reply doc/source/rllib/single-agent-episode.rst:27]: Yeah, but also added this to the actual text.

File: doc/source/rllib/single-agent-episode.rst (hunk 3 above)
Comment (sven1977) [reply doc/source/rllib/single-agent-episode.rst:93]: Added more context and a link to the `get_extra_model_outputs` docstring.

File: doc/source/rllib/single-agent-episode.rst (hunk 4 above)
Comment (sven1977) [reply doc/source/rllib/single-agent-episode.rst:111]: I think "to these" is ok here.

File: doc/source/rllib/single-agent-episode.rst (hunk 5 above)
Comment (sven1977) [reply doc/source/rllib/single-agent-episode.rst:132]: fixed

File: doc/source/rllib/single-agent-episode.rst (hunk 6 above)
Comment (sven1977) [reply doc/source/rllib/single-agent-episode.rst:134]: fixed

File: doc/source/rllib/single-agent-episode.rst (hunk 7 above)
Comment (sven1977) [reply doc/source/rllib/single-agent-episode.rst:149]: fixed

File: doc/source/rllib/single-agent-episode.rst (hunk 8 above)
Comment (sven1977) [reply doc/source/rllib/single-agent-episode.rst:211]: I think this is fine.

Note that we do not support indexing by global (actual gym episode) timestep yet. It's probably a usecase that's never really needed.

So if you have an episode chunk from global timestep 100 to global timestep 200, as a user, you would normally just loop through it (from `local_index` 0 to `local_index` 100) and then - at each `local_index` - say: I want to do 4x-framestacking from `local_index - 3` to `local_index + 1`. This means that if local_index is 0 or 1 or 2, you'd have a negative start index and for these few cases you want to tell the getter that these negative indices indeed mean "go into the lookback buffer" instead of "count from the end".

In other words: If you want to index into the lookback buffer, you have to provide a negative number, b/c `index=0` is always the first item after(!) the lookback buffer.
